thisisarchimedes/Archimedes_Finance
# AccessController

Two main motivations for doing this are:
- Consolidate AccessControl logic, initialization, and reentrancy modifier in one place
- Make it easy for tests to change their roles to test different behavior
  - For example to test some PositionToken functions we need executive to be owner but in ContractTestContext it is LeverageEngine. ContractTestContext should always represent the final expected deployed state by default.

In detail:
- Add new AccessController abstract base class to better share and standardize common methods around access control
  - admin set in constructor to contract deployer
  - new setRoles function for executive, governor, guardian with modifiers only[Role] for each
  - Allow address getters for each role internally only so that child contracts can grant access to needed addresses
  - Simple _init function to set initialized = true for contracts that use expectInitialized. Possibly additional base initialization logic can be added here if needed in future.
- Allow tests to override role addresses in argument to buildContractTestContext
- Modify buildContractTestContext:
  - New `deployContracts` to build and deploy contracts concurrently that provides admin address as first argument to constructor. Additional constructor args can still be provided.
  - Call new setRoles on all contracts to setup roles consistently for all contracts
- Remove any lone `deploy` functions from other tests to ensure the contracts are constructed properly from buildContractTestContext.
